File: app/agent/src/models/knowledge.py
"""Knowledge base related SQLAlchemy models based on existing migrations."""

from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy import String, Text, Integer, Boolean, Float, DateTime, ForeignKey, BIGINT
from sqlalchemy.dialects.postgresql import UUID, JSONB
from pgvector.sqlalchemy import Vector
from uuid import uuid4
from datetime import datetime
from typing import Optional, Dict, Any, List

from .base import BaseModel


class KnowledgeDocument(BaseModel):
    """Knowledge document model from migration 022."""
    
    __tablename__ = "knowledge_documents"
    
    id: Mapped[UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True), 
        ForeignKey("tenants.id", ondelete="CASCADE"),
        nullable=False
    )
    project_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False
    )
    filename: Mapped[str] = mapped_column(String(255), nullable=False)
    content_type: Mapped[str] = mapped_column(String(100), nullable=False)
    file_size: Mapped[int] = mapped_column(BIGINT, nullable=False)
    file_path: Mapped[str] = mapped_column(Text, nullable=False)
    original_content: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    processed_content: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="processing")
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    metadata: Mapped[Dict[str, Any]] = mapped_column(JSONB, default=dict)
    
    # Relationships
    chunks: Mapped[List["KnowledgeChunk"]] = relationship(
        "KnowledgeChunk",
        back_populates="document",
        cascade="all, delete-orphan"
    )


class KnowledgeChunk(BaseModel):
    """Knowledge chunk model for embeddings from migration 022."""
    
    __tablename__ = "knowledge_chunks"
    
    id: Mapped[UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    document_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("knowledge_documents.id", ondelete="CASCADE"),
        nullable=False
    )
    chunk_index: Mapped[int] = mapped_column(Integer, nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    token_count: Mapped[int] = mapped_column(Integer, nullable=False)
    embedding: Mapped[Optional[List[float]]] = mapped_column(Vector(1536), nullable=True)  # Made nullable in migration 023
    metadata: Mapped[Dict[str, Any]] = mapped_column(JSONB, default=dict)
    
    # Relationships
    document: Mapped["KnowledgeDocument"] = relationship("KnowledgeDocument", back_populates="chunks")


class KnowledgeScrapingJob(BaseModel):
    """Web scraping job model from migration 022."""
    
    __tablename__ = "knowledge_scraping_jobs"
    
    id: Mapped[UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("tenants.id", ondelete="CASCADE"),
        nullable=False
    )
    project_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False
    )
    url: Mapped[str] = mapped_column(Text, nullable=False)
    max_depth: Mapped[int] = mapped_column(Integer, nullable=False, default=5)
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="pending")
    pages_scraped: Mapped[int] = mapped_column(Integer, default=0)
    total_pages: Mapped[int] = mapped_column(Integer, default=0)
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    started_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    
    # Relationships
    pages: Mapped[List["KnowledgeScrapedPage"]] = relationship(
        "KnowledgeScrapedPage",
        back_populates="job",
        cascade="all, delete-orphan"
    )


class KnowledgeScrapedPage(BaseModel):
    """Scraped page model from migration 022."""
    
    __tablename__ = "knowledge_scraped_pages"
    
    id: Mapped[UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    job_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("knowledge_scraping_jobs.id", ondelete="CASCADE"),
        nullable=False
    )
    url: Mapped[str] = mapped_column(Text, nullable=False)
    title: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    token_count: Mapped[int] = mapped_column(Integer, nullable=False)
    scraped_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now())
    embedding: Mapped[Optional[List[float]]] = mapped_column(Vector(1536), nullable=True)  # Made nullable in migration 023
    metadata: Mapped[Dict[str, Any]] = mapped_column(JSONB, default=dict)
    content_hash: Mapped[Optional[str]] = mapped_column(String(64), nullable=True)  # Added in migration 024/026
    page_id: Mapped[Optional[UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("knowledge_pages.id", ondelete="SET NULL"),
        nullable=True
    )  # Added in migration 025
    
    # Relationships
    job: Mapped["KnowledgeScrapingJob"] = relationship("KnowledgeScrapingJob", back_populates="pages")
    page: Mapped[Optional["KnowledgePage"]] = relationship("KnowledgePage", back_populates="scraped_versions")


class KnowledgePage(BaseModel):
    """Deduplicated knowledge page model from migration 025."""
    
    __tablename__ = "knowledge_pages"
    
    id: Mapped[UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("tenants.id", ondelete="CASCADE"),
        nullable=False
    )
    project_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False
    )
    url: Mapped[str] = mapped_column(Text, nullable=False)
    content_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    title: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    token_count: Mapped[int] = mapped_column(Integer, nullable=False)
    embedding: Mapped[Optional[List[float]]] = mapped_column(Vector(1536), nullable=True)
    metadata: Mapped[Dict[str, Any]] = mapped_column(JSONB, default=dict)
    first_scraped_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now())
    last_scraped_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now())
    scrape_count: Mapped[int] = mapped_column(Integer, default=1)
    
    # Relationships
    scraped_versions: Mapped[List["KnowledgeScrapedPage"]] = relationship(
        "KnowledgeScrapedPage",
        back_populates="page"
    )


class KnowledgeSettings(BaseModel):
    """Knowledge base settings per project from migration 022."""
    
    __tablename__ = "knowledge_settings"
    
    id: Mapped[UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("tenants.id", ondelete="CASCADE"),
        nullable=False
    )
    project_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False
    )
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    embedding_model: Mapped[str] = mapped_column(String(100), nullable=False, default="text-embedding-ada-002")
    chunk_size: Mapped[int] = mapped_column(Integer, nullable=False, default=1000)
    chunk_overlap: Mapped[int] = mapped_column(Integer, nullable=False, default=200)
    max_context_chunks: Mapped[int] = mapped_column(Integer, nullable=False, default=5)
    similarity_threshold: Mapped[float] = mapped_column(Float, nullable=False, default=0.7)

File: app/backend/internal/repo/knowledge.go
package repo

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/jmoiron/sqlx"
	"github.com/lib/pq"
	"github.com/pgvector/pgvector-go"

	"github.com/bareuptime/tms/internal/models"
)

type KnowledgeRepository struct {
	db *sqlx.DB
}

func NewKnowledgeRepository(db *sqlx.DB) *KnowledgeRepository {
	return &KnowledgeRepository{db: db}
}

// generateContentHash creates a SHA256 hash of the content for deduplication
func (r *KnowledgeRepository) generateContentHash(content string) string {
	hash := sha256.Sum256([]byte(content))
	return fmt.Sprintf("%x", hash)
}

// Document operations

func (r *KnowledgeRepository) CreateDocument(doc *models.KnowledgeDocument) error {
	query := `
		INSERT INTO knowledge_documents (
			id, tenant_id, project_id, filename, content_type, file_size, 
			file_path, original_content, processed_content, status, 
			error_message, metadata
		) VALUES (
			:id, :tenant_id, :project_id, :filename, :content_type, :file_size,
			:file_path, :original_content, :processed_content, :status,
			:error_message, :metadata
		)`

	_, err := r.db.NamedExec(query, doc)
	return err
}

func (r *KnowledgeRepository) GetDocument(id uuid.UUID) (*models.KnowledgeDocument, error) {
	var doc models.KnowledgeDocument
	query := `SELECT * FROM knowledge_documents WHERE id = $1`

	err := r.db.Get(&doc, query, id)
	if err != nil {
		return nil, err
	}

	return &doc, nil
}

func (r *KnowledgeRepository) ListDocuments(tenantID, projectID uuid.UUID, limit, offset int) ([]*models.KnowledgeDocument, error) {
	var docs []*models.KnowledgeDocument
	query := `
		SELECT * FROM knowledge_documents 
		WHERE tenant_id = $1 AND project_id = $2
		ORDER BY created_at DESC
		LIMIT $3 OFFSET $4`

	err := r.db.Select(&docs, query, tenantID, projectID, limit, offset)
	return docs, err
}

func (r *KnowledgeRepository) UpdateDocumentStatus(id uuid.UUID, status string, errorMessage *string) error {
	query := `
		UPDATE knowledge_documents 
		SET status = $2, error_message = $3, updated_at = NOW()
		WHERE id = $1`

	_, err := r.db.Exec(query, id, status, errorMessage)
	return err
}

func (r *KnowledgeRepository) UpdateDocumentContent(id uuid.UUID, processedContent string) error {
	query := `
		UPDATE knowledge_documents 
		SET processed_content = $2, updated_at = NOW()
		WHERE id = $1`

	_, err := r.db.Exec(query, id, processedContent)
	return err
}

func (r *KnowledgeRepository) DeleteDocument(id uuid.UUID) error {
	query := `DELETE FROM knowledge_documents WHERE id = $1`
	_, err := r.db.Exec(query, id)
	return err
}

// Chunk operations

func (r *KnowledgeRepository) CreateChunk(chunk *models.KnowledgeChunk) error {
	query := `
		INSERT INTO knowledge_chunks (
			id, document_id, chunk_index, content, token_count, embedding, metadata
		) VALUES (
			:id, :document_id, :chunk_index, :content, :token_count, :embedding, :metadata
		)`

	_, err := r.db.NamedExec(query, chunk)
	return err
}

func (r *KnowledgeRepository) CreateChunks(chunks []*models.KnowledgeChunk) error {
	if len(chunks) == 0 {
		return nil
	}

	tx, err := r.db.Beginx()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	query := `
		INSERT INTO knowledge_chunks (
			id, document_id, chunk_index, content, token_count, embedding, metadata
		) VALUES (
			:id, :document_id, :chunk_index, :content, :token_count, :embedding, :metadata
		)`

	for _, chunk := range chunks {
		_, err := tx.NamedExec(query, chunk)
		if err != nil {
			return err
		}
	}

	return tx.Commit()
}

func (r *KnowledgeRepository) GetDocumentChunks(documentID uuid.UUID) ([]*models.KnowledgeChunk, error) {
	var chunks []*models.KnowledgeChunk
	query := `
		SELECT * FROM knowledge_chunks 
		WHERE document_id = $1 
		ORDER BY chunk_index`

	err := r.db.Select(&chunks, query, documentID)
	return chunks, err
}

func (r *KnowledgeRepository) SearchSimilarChunks(tenantID, projectID uuid.UUID, embedding pgvector.Vector, limit int, threshold float64) ([]*models.KnowledgeSearchResult, error) {
	query := `
		SELECT 
			kc.id,
			'document' as type,
			kc.content,
			1 - (kc.embedding <=> $1) as score,
			kd.filename as source,
			NULL as title,
			kd.id as document_id,
			NULL as job_id,
			kc.chunk_index,
			kc.metadata
		FROM knowledge_chunks kc
		JOIN knowledge_documents kd ON kc.document_id = kd.id
		WHERE kd.tenant_id = $2 AND kd.project_id = $3 
		AND kd.status = 'completed'
		AND kc.embedding IS NOT NULL
		AND 1 - (kc.embedding <=> $1) > $4
		ORDER BY kc.embedding <=> $1
		LIMIT $5`

	var results []*models.KnowledgeSearchResult
	err := r.db.Select(&results, query, embedding, tenantID, projectID, threshold, limit)
	return results, err
}

// Scraping job operations

func (r *KnowledgeRepository) CreateScrapingJob(job *models.KnowledgeScrapingJob) error {
	query := `
		INSERT INTO knowledge_scraping_jobs (
			id, tenant_id, project_id, url, max_depth, status
		) VALUES (
			:id, :tenant_id, :project_id, :url, :max_depth, :status
		)`

	_, err := r.db.NamedExec(query, job)
	return err
}

func (r *KnowledgeRepository) GetScrapingJob(id, tenantID, projectID uuid.UUID) (*models.KnowledgeScrapingJob, error) {
	var job models.KnowledgeScrapingJob
	query := `SELECT * FROM knowledge_scraping_jobs WHERE id = $1 aND tenant_id = $2 AND project_id = $3`

	err := r.db.Get(&job, query, id, tenantID, projectID)
	if err != nil {
		return nil, err
	}

	return &job, nil
}

func (r *KnowledgeRepository) ListScrapingJobs(tenantID, projectID uuid.UUID, limit, offset int) ([]*models.KnowledgeScrapingJob, error) {
	var jobs []*models.KnowledgeScrapingJob
	query := `
		SELECT * FROM knowledge_scraping_jobs 
		WHERE tenant_id = $1 AND project_id = $2
		ORDER BY created_at DESC
		LIMIT $3 OFFSET $4`

	err := r.db.Select(&jobs, query, tenantID, projectID, limit, offset)
	return jobs, err
}

func (r *KnowledgeRepository) UpdateScrapingJobStatus(id uuid.UUID, status string, errorMessage *string) error {
	query := `
		UPDATE knowledge_scraping_jobs 
		SET status = $2, error_message = $3, updated_at = NOW()
		WHERE id = $1`

	_, err := r.db.Exec(query, id, status, errorMessage)
	return err
}

func (r *KnowledgeRepository) UpdateScrapingJobProgress(id uuid.UUID, pagesScraped, totalPages int) error {
	query := `
		UPDATE knowledge_scraping_jobs 
		SET pages_scraped = $2, total_pages = $3, updated_at = NOW()
		WHERE id = $1`

	_, err := r.db.Exec(query, id, pagesScraped, totalPages)
	return err
}

func (r *KnowledgeRepository) StartScrapingJob(id uuid.UUID) error {
	query := `
		UPDATE knowledge_scraping_jobs 
		SET status = 'running', started_at = NOW(), updated_at = NOW()
		WHERE id = $1`

	_, err := r.db.Exec(query, id)
	return err
}

func (r *KnowledgeRepository) CompleteScrapingJob(id uuid.UUID) error {
	query := `
        UPDATE knowledge_scraping_jobs 
        SET status = 'completed', completed_at = NOW(), updated_at = NOW()
        WHERE id = $1`

	_, err := r.db.Exec(query, id)
	return err
}

// MarkJobAwaitingSelection stores the staging file and moves the job into the selection phase
func (r *KnowledgeRepository) MarkJobAwaitingSelection(id uuid.UUID, totalPages int, stagingFilePath string) error {
	query := `
        UPDATE knowledge_scraping_jobs
        SET status = 'awaiting_selection',
            pages_scraped = $2,
            total_pages = $2,
            staging_file_path = $3,
            selected_links = '{}'::text[],
            indexing_started_at = NULL,
            indexing_completed_at = NULL,
            completed_at = NULL,
            updated_at = NOW()
        WHERE id = $1`

	_, err := r.db.Exec(query, id, totalPages, stagingFilePath)
	return err
}

// StartIndexingJob records the selected links and marks the indexing phase as started
func (r *KnowledgeRepository) StartIndexingJob(id uuid.UUID, selectedLinks []string) error {
	query := `
        UPDATE knowledge_scraping_jobs
        SET status = 'indexing',
            selected_links = $2,
            indexing_started_at = NOW(),
            indexing_completed_at = NULL,
            updated_at = NOW()
        WHERE id = $1`

	_, err := r.db.Exec(query, id, pq.Array(selectedLinks))
	return err
}

// CompleteIndexingJob finalises the job and sets completion timestamps
func (r *KnowledgeRepository) CompleteIndexingJob(id uuid.UUID) error {
	query := `
        UPDATE knowledge_scraping_jobs
        SET status = 'completed',
            indexing_completed_at = NOW(),
            completed_at = NOW(),
            updated_at = NOW()
        WHERE id = $1`

	_, err := r.db.Exec(query, id)
	return err
}

// SaveSelectedLinks stores the set of links chosen by the user without altering job status
func (r *KnowledgeRepository) SaveSelectedLinks(id uuid.UUID, selectedLinks []string) error {
	query := `
        UPDATE knowledge_scraping_jobs
        SET selected_links = $2,
            updated_at = NOW()
        WHERE id = $1`

	_, err := r.db.Exec(query, id, pq.Array(selectedLinks))
	return err
}

// Scraped page operations

func (r *KnowledgeRepository) CreateScrapedPage(page *models.KnowledgeScrapedPage) error {
	query := `
		INSERT INTO knowledge_scraped_pages (
			id, job_id, url, title, content, content_hash, token_count, embedding, metadata
		) VALUES (
			:id, :job_id, :url, :title, :content, :content_hash, :token_count, :embedding, :metadata
		)`

	_, err := r.db.NamedExec(query, page)
	return err
}

func (r *KnowledgeRepository) CreateScrapedPages(pages []*models.KnowledgeScrapedPage) error {
	fmt.Println("Creating scraped pages with content-aware deduplication...")
	if len(pages) == 0 {
		return nil
	}

	tx, err := r.db.Beginx()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	// First, get the project_id from the job
	if len(pages) > 0 {
		var projectID uuid.UUID
		err := tx.Get(&projectID,
			"SELECT project_id FROM knowledge_scraping_jobs WHERE id = $1",
			pages[0].JobID)
		if err != nil {
			return fmt.Errorf("failed to get project_id: %w", err)
		}

		// Get existing URLs and their content hashes for this project
		type ExistingPage struct {
			URL         string    `db:"url"`
			ContentHash *string   `db:"content_hash"`
			ID          uuid.UUID `db:"id"`
		}

		var existingPages []ExistingPage
		err = tx.Select(&existingPages, `
			SELECT DISTINCT ksp.id, ksp.url, ksp.content_hash
			FROM knowledge_scraped_pages ksp 
			JOIN knowledge_scraping_jobs ksj ON ksp.job_id = ksj.id 
			WHERE ksj.project_id = $1`, projectID)
		if err != nil {
			return fmt.Errorf("failed to get existing pages: %w", err)
		}

		// Create maps for fast lookup
		existingURLs := make(map[string]ExistingPage)
		for _, page := range existingPages {
			existingURLs[page.URL] = page
		}

		// Process each page: new, updated, or duplicate
		var newPages []*models.KnowledgeScrapedPage
		var updatedPages []*models.KnowledgeScrapedPage
		duplicateCount := 0
		updatedCount := 0

		for _, page := range pages {
			existing, urlExists := existingURLs[page.URL]

			if !urlExists {
				// Completely new URL - needs embedding
				newPages = append(newPages, page)
				// Mark that this page needs an embedding by keeping embedding nil
				page.Embedding = nil
			} else if existing.ContentHash != nil && page.ContentHash != nil &&
				*existing.ContentHash != *page.ContentHash {
				// URL exists but content changed - this is an update, needs new embedding
				page.ID = existing.ID // Keep same ID for update
				updatedPages = append(updatedPages, page)
				updatedCount++
				// Mark that this page needs an embedding by keeping embedding nil
				page.Embedding = nil
			} else {
				// Same URL and same content - skip, no embedding needed
				duplicateCount++
				// Mark that this page does NOT need an embedding
				page.ID = existing.ID               // Set the existing ID
				page.Embedding = &pgvector.Vector{} // Dummy value to indicate no embedding needed
			}
		}

		fmt.Printf("Analysis: %d new pages, %d updated pages, %d duplicates skipped\n",
			len(newPages), len(updatedPages), duplicateCount)

		// Insert new pages
		if len(newPages) > 0 {
			insertQuery := `
				INSERT INTO knowledge_scraped_pages (
					id, job_id, url, title, content, content_hash, token_count, embedding, metadata
				) VALUES (
					:id, :job_id, :url, :title, :content, :content_hash, :token_count, :embedding, :metadata
				)`

			for _, page := range newPages {
				_, err := tx.NamedExec(insertQuery, page)
				if err != nil {
					return fmt.Errorf("failed to insert new page: %w", err)
				}
			}
		}

		// Update existing pages with new content
		if len(updatedPages) > 0 {
			updateQuery := `
				UPDATE knowledge_scraped_pages 
				SET title = :title, content = :content, content_hash = :content_hash, 
				    token_count = :token_count, embedding = :embedding, 
				    metadata = :metadata, scraped_at = NOW()
				WHERE id = :id`

			for _, page := range updatedPages {
				_, err := tx.NamedExec(updateQuery, page)
				if err != nil {
					return fmt.Errorf("failed to update page: %w", err)
				}
			}
		}

		fmt.Printf("Successfully processed: %d inserted, %d updated, %d skipped\n",
			len(newPages), len(updatedPages), duplicateCount)
	}

	return tx.Commit()
}

// UpdatePageEmbeddings updates the embeddings for a list of pages
func (r *KnowledgeRepository) UpdatePageEmbeddings(pages []*models.KnowledgeScrapedPage) error {
	if len(pages) == 0 {
		return nil
	}

	tx, err := r.db.Beginx()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	updateQuery := `UPDATE knowledge_scraped_pages SET embedding = $2 WHERE id = $1`

	for _, page := range pages {
		if page.Embedding != nil && page.ID != uuid.Nil {
			_, err := tx.Exec(updateQuery, page.ID, page.Embedding)
			if err != nil {
				return fmt.Errorf("failed to update embedding for page %s: %w", page.ID, err)
			}
		}
	}

	return tx.Commit()
}

func (r *KnowledgeRepository) GetJobPages(jobID, tenantID, projectID uuid.UUID) ([]*models.KnowledgeScrapedPage, error) {
	var pages []*models.KnowledgeScrapedPage
	query := `
		SELECT * FROM knowledge_scraped_pages 
		WHERE job_id = $1 and tenant_id = $2 and project_id = $3
		ORDER BY scraped_at`

	err := r.db.Select(&pages, query, jobID, tenantID, projectID)
	return pages, err
}

func (r *KnowledgeRepository) SearchSimilarPages(tenantID, projectID uuid.UUID, embedding pgvector.Vector, limit int, threshold float64) ([]*models.KnowledgeSearchResult, error) {
	query := `
		SELECT 
			ksp.id,
			'webpage' as type,
			ksp.content,
			1 - (ksp.embedding <=> $1) as score,
			ksp.url as source,
			ksp.title,
			NULL as document_id,
			ksp.job_id,
			NULL as chunk_index,
			ksp.metadata
		FROM knowledge_scraped_pages ksp
		JOIN knowledge_scraping_jobs ksj ON ksp.job_id = ksj.id
		WHERE ksj.tenant_id = $2 AND ksj.project_id = $3 
		AND ksj.status = 'completed'
		AND ksp.embedding IS NOT NULL
		AND 1 - (ksp.embedding <=> $1) > $4
		ORDER BY ksp.embedding <=> $1
		LIMIT $5`

	var results []*models.KnowledgeSearchResult
	err := r.db.Select(&results, query, embedding, tenantID, projectID, threshold, limit)
	return results, err
}

// Combined search across chunks and pages
func (r *KnowledgeRepository) SearchKnowledgeBase(tenantID, projectID uuid.UUID, embedding pgvector.Vector, limit int, threshold float64, includeDocuments, includePages bool) ([]*models.KnowledgeSearchResult, error) {
	var results []*models.KnowledgeSearchResult

	if includeDocuments {
		chunkResults, err := r.SearchSimilarChunks(tenantID, projectID, embedding, limit, threshold)
		if err != nil {
			return nil, err
		}
		results = append(results, chunkResults...)
	}

	if includePages {
		pageResults, err := r.SearchSimilarPages(tenantID, projectID, embedding, limit, threshold)
		if err != nil {
			return nil, err
		}
		results = append(results, pageResults...)
	}

	// Sort by score descending and limit results
	if len(results) > limit {
		// Sort by score
		for i := 0; i < len(results)-1; i++ {
			for j := 0; j < len(results)-i-1; j++ {
				if results[j].Score < results[j+1].Score {
					results[j], results[j+1] = results[j+1], results[j]
				}
			}
		}
		results = results[:limit]
	}

	return results, nil
}

// Settings operations

func (r *KnowledgeRepository) GetSettings(projectID uuid.UUID) (*models.KnowledgeSettings, error) {
	var settings models.KnowledgeSettings
	query := `SELECT * FROM knowledge_settings WHERE project_id = $1`

	err := r.db.Get(&settings, query, projectID)
	if err != nil {
		return nil, err
	}

	return &settings, nil
}

func (r *KnowledgeRepository) CreateSettings(settings *models.KnowledgeSettings) error {
	query := `
		INSERT INTO knowledge_settings (
			id, tenant_id, project_id, enabled, embedding_model, chunk_size,
			chunk_overlap, max_context_chunks, similarity_threshold
		) VALUES (
			:id, :tenant_id, :project_id, :enabled, :embedding_model, :chunk_size,
			:chunk_overlap, :max_context_chunks, :similarity_threshold
		)`

	_, err := r.db.NamedExec(query, settings)
	return err
}

func (r *KnowledgeRepository) UpdateSettings(projectID uuid.UUID, updates *models.UpdateKnowledgeSettingsRequest) error {
	setParts := []string{}
	args := []interface{}{}
	argIndex := 1

	if updates.Enabled != nil {
		setParts = append(setParts, fmt.Sprintf("enabled = $%d", argIndex))
		args = append(args, *updates.Enabled)
		argIndex++
	}

	if updates.EmbeddingModel != nil {
		setParts = append(setParts, fmt.Sprintf("embedding_model = $%d", argIndex))
		args = append(args, *updates.EmbeddingModel)
		argIndex++
	}

	if updates.ChunkSize != nil {
		setParts = append(setParts, fmt.Sprintf("chunk_size = $%d", argIndex))
		args = append(args, *updates.ChunkSize)
		argIndex++
	}

	if updates.ChunkOverlap != nil {
		setParts = append(setParts, fmt.Sprintf("chunk_overlap = $%d", argIndex))
		args = append(args, *updates.ChunkOverlap)
		argIndex++
	}

	if updates.MaxContextChunks != nil {
		setParts = append(setParts, fmt.Sprintf("max_context_chunks = $%d", argIndex))
		args = append(args, *updates.MaxContextChunks)
		argIndex++
	}

	if updates.SimilarityThreshold != nil {
		setParts = append(setParts, fmt.Sprintf("similarity_threshold = $%d", argIndex))
		args = append(args, *updates.SimilarityThreshold)
		argIndex++
	}

	if len(setParts) == 0 {
		return nil // No updates
	}

	setParts = append(setParts, fmt.Sprintf("updated_at = $%d", argIndex))
	args = append(args, time.Now())
	argIndex++

	query := fmt.Sprintf("UPDATE knowledge_settings SET %s WHERE project_id = $%d",
		fmt.Sprintf("%s", setParts), argIndex)
	args = append(args, projectID)

	_, err := r.db.Exec(query, args...)
	return err
}

// Statistics operations

func (r *KnowledgeRepository) GetStats(tenantID, projectID uuid.UUID) (*models.KnowledgeStats, error) {
	stats := &models.KnowledgeStats{}

	// Count documents
	err := r.db.Get(&stats.TotalDocuments,
		"SELECT COUNT(*) FROM knowledge_documents WHERE tenant_id = $1 AND project_id = $2",
		tenantID, projectID)
	if err != nil && err != sql.ErrNoRows {
		return nil, err
	}

	// Count chunks
	err = r.db.Get(&stats.TotalChunks,
		`SELECT COUNT(*) FROM knowledge_chunks kc 
		 JOIN knowledge_documents kd ON kc.document_id = kd.id 
		 WHERE kd.tenant_id = $1 AND kd.project_id = $2`,
		tenantID, projectID)
	if err != nil && err != sql.ErrNoRows {
		return nil, err
	}

	// Count scraping jobs
	err = r.db.Get(&stats.TotalScrapingJobs,
		"SELECT COUNT(*) FROM knowledge_scraping_jobs WHERE tenant_id = $1 AND project_id = $2",
		tenantID, projectID)
	if err != nil && err != sql.ErrNoRows {
		return nil, err
	}

	// Count scraped pages
	err = r.db.Get(&stats.TotalScrapedPages,
		`SELECT COUNT(*) FROM knowledge_scraped_pages ksp 
		 JOIN knowledge_scraping_jobs ksj ON ksp.job_id = ksj.id 
		 WHERE ksj.tenant_id = $1 AND ksj.project_id = $2`,
		tenantID, projectID)
	if err != nil && err != sql.ErrNoRows {
		return nil, err
	}

	// Calculate total storage
	err = r.db.Get(&stats.TotalStorageBytes,
		"SELECT COALESCE(SUM(file_size), 0) FROM knowledge_documents WHERE tenant_id = $1 AND project_id = $2",
		tenantID, projectID)
	if err != nil && err != sql.ErrNoRows {
		return nil, err
	}

	return stats, nil
}

// ReplaceFAQItems replaces all FAQ entries for a given tenant/project with the provided items
func (r *KnowledgeRepository) ReplaceFAQItems(ctx context.Context, tenantID, projectID uuid.UUID, items []*models.KnowledgeFAQItem) error {
	tx, err := r.db.BeginTxx(ctx, nil)
	if err != nil {
		return err
	}
	defer tx.Rollback()

	if _, err := tx.ExecContext(ctx,
		"DELETE FROM knowledge_faq_items WHERE tenant_id = $1 AND project_id = $2",
		tenantID, projectID,
	); err != nil {
		return fmt.Errorf("failed to clear existing FAQ items: %w", err)
	}

	insertQuery := `
		INSERT INTO knowledge_faq_items (
			id, tenant_id, project_id, question, answer, source_url, source_section, metadata, created_at, updated_at
		) VALUES (
			:id, :tenant_id, :project_id, :question, :answer, :source_url, :source_section, :metadata, :created_at, :updated_at
		)`

	for _, item := range items {
		if item.Metadata == nil {
			item.Metadata = models.JSONMap{}
		}
		if item.CreatedAt.IsZero() {
			item.CreatedAt = time.Now()
		}
		if item.UpdatedAt.IsZero() {
			item.UpdatedAt = item.CreatedAt
		}
		if _, err := tx.NamedExecContext(ctx, insertQuery, item); err != nil {
			return fmt.Errorf("failed to insert FAQ item: %w", err)
		}
	}

	return tx.Commit()
}

// ListFAQItems returns FAQ entries for a tenant/project ordered by creation time
func (r *KnowledgeRepository) ListFAQItems(ctx context.Context, tenantID, projectID uuid.UUID) ([]*models.KnowledgeFAQItem, error) {
	items := []*models.KnowledgeFAQItem{}
	query := `
		SELECT * FROM knowledge_faq_items
		WHERE tenant_id = $1 AND project_id = $2
		ORDER BY created_at ASC`

	if err := r.db.SelectContext(ctx, &items, query, tenantID, projectID); err != nil {
		if err == sql.ErrNoRows {
			return []*models.KnowledgeFAQItem{}, nil
		}
		return nil, err
	}

	return items, nil
}
